distributed/tests/test_worker.py

```diff
@@ -3502,3 +3505,43 @@ async def test_reconnect_argument_deprecated(s):
         warnings.simplefilter("error")
         async with Worker(s.address):
             pass
+
+
+@gen_test()
+async def test_worker_makes_own_thread():
+    class InitWorkerNewThread(WorkerPlugin):
+        name: name
+        thread: threading.Thread
+
+        def __init__(self, scheduler):
+            super().__init__()
+            self.scheduler = scheduler
+
+        def setup(self, worker):
+            self.loop = asyncio.new_event_loop()
+            self.thread = threading.Thread(target=self.loop.run_forever)
```

Comment:
I don't understand the call to loop.run_forever here, after you use run_coroutine_threadsafe you're not running any further tasks on the loop, so you should only need `asyncio.run(func())`

In addition this is a synchronous setup method which means you're blocking the worker thread while calling `await client.register_worker_plugin`, so something like `await to_thread(asyncio.run, func())` is needed

```python
@gen_test()
async def test_worker_makes_own_thread():
    class InitWorkerNewThread(WorkerPlugin):
        name: name
        thread: threading.Thread

        def __init__(self, scheduler):
            super().__init__()
            self.scheduler = scheduler

        async def setup(self, worker):
            async def func():
                try:
                    worker = get_worker()
                except ValueError:
                    async with Client(self.scheduler, asynchronous=True) as client:
                        await client.register_worker_plugin(
                            InitWorkerNewThread(self.scheduler)
                        )

            await to_thread(asyncio.run, func())

        def teardown(self, worker):
            pass
```

Reply:
I'll look into this, as I'm not totally sure. 

The code I'm working on was actually written by @crusaderky in his previous job though, so he might be able to explain (though he wrote this code 2 years ago, so it might stretch his memory a bit).